nexus/src/app/sagas/region_replacement_start.rs

```diff
@@ -0,0 +1,1261 @@
+
+#[cfg(test)]
+pub(crate) mod test {
+    use crate::{
+        app::db::lookup::LookupPath, app::db::DataStore,
+        app::saga::create_saga_dag,
+        app::sagas::region_replacement_start::find_only_new_region,
+        app::sagas::region_replacement_start::Params,
+        app::sagas::region_replacement_start::SagaRegionReplacementStart,
+        app::sagas::test_helpers::test_opctx, app::RegionAllocationStrategy,
+    };
+    use chrono::Utc;
+    use nexus_db_model::Dataset;
+    use nexus_db_model::DatasetKind;
+    use nexus_db_model::Region;
+    use nexus_db_model::RegionReplacement;
+    use nexus_db_model::RegionReplacementState;
+    use nexus_db_model::Volume;
+    use nexus_db_queries::authn::saga::Serialized;
+    use nexus_db_queries::context::OpContext;
+    use nexus_test_utils::resource_helpers::create_disk;
+    use nexus_test_utils::resource_helpers::create_project;
+    use nexus_test_utils::resource_helpers::DiskTest;
+    use nexus_test_utils_macros::nexus_test;
+    use nexus_types::identity::Asset;
+    use sled_agent_client::types::VolumeConstructionRequest;
+    use uuid::Uuid;
+
+    type ControlPlaneTestContext =
+        nexus_test_utils::ControlPlaneTestContext<crate::Server>;
+
+    const DISK_NAME: &str = "my-disk";
+    const PROJECT_NAME: &str = "springfield-squidport";
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_region_replacement_start_saga(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let mut disk_test = DiskTest::new(cptestctx).await;
+        disk_test.add_zpool_with_dataset(&cptestctx).await;
+
+        let client = &cptestctx.external_client;
+        let nexus = &cptestctx.server.server_context().nexus;
+        let datastore = nexus.datastore();
+        let opctx = test_opctx(cptestctx);
+
+        let _project_id =
+            create_project(&client, PROJECT_NAME).await.identity.id;
+
+        // Create a disk
+        let client = &cptestctx.external_client;
+        let disk = create_disk(&client, PROJECT_NAME, DISK_NAME).await;
+
+        // Assert disk has three allocated regions
+        let disk_id = disk.identity.id;
+        let (.., db_disk) = LookupPath::new(&opctx, &datastore)
+            .disk_id(disk_id)
+            .fetch()
+            .await
+            .unwrap_or_else(|_| panic!("test disk {:?} should exist", disk_id));
+
+        let allocated_regions =
+            datastore.get_allocated_regions(db_disk.volume_id).await.unwrap();
+        assert_eq!(allocated_regions.len(), 3);
+
+        // Replace one of the disk's regions
+        let region_to_replace: &nexus_db_model::Region =
+            &allocated_regions[0].1;
+
+        // Manually insert the replacement request
+        let request = RegionReplacement {
+            id: Uuid::new_v4(),
+            request_time: Utc::now(),
+            old_region_id: region_to_replace.id(),
+            volume_id: region_to_replace.volume_id(),
+            old_region_volume_id: None,
```

Comment:
Without changing the schema, this is a little awkward. I think this would work:

```rust
diff --git a/nexus/db-model/src/region_replacement.rs b/nexus/db-model/src/region_replacement.rs
index a04710f53..3984581f5 100644
--- a/nexus/db-model/src/region_replacement.rs
+++ b/nexus/db-model/src/region_replacement.rs
@@ -135,7 +135,8 @@ pub struct RegionReplacement {
     pub volume_id: Uuid,
 
     /// A synthetic volume that only is used to later delete the old region
-    pub old_region_volume_id: Option<Uuid>,
+    #[diesel(column_name = old_region_volume_id)]
+    pub delete_old_region_volume_id: Option<Uuid>,
 
     /// The new region that will be used to replace the old one
     pub new_region_id: Option<Uuid>,
```

but without changing `schema.rs` as well would still lead to code referencing both:

```rust
        use db::schema::region_replacement::dsl;
        let updated = diesel::update(dsl::region_replacement)
            .filter(dsl::id.eq(region_replacement_id))
            .filter(dsl::operating_saga_id.eq(operating_saga_id))
            .filter(
                dsl::replacement_state.eq(RegionReplacementState::Allocating),
            )
            .set((
                dsl::replacement_state.eq(RegionReplacementState::Running),
                dsl::old_region_volume_id.eq(Some(old_region_volume_id)),
                dsl::new_region_id.eq(Some(new_region_id)),
                dsl::operating_saga_id.eq(Option::<Uuid>::None),
            ))
            .check_if_exists::<RegionReplacement>(region_replacement_id)
            .execute_and_check(&*self.pool_connection_authorized(opctx).await?)
            .await;
    
        match updated {
            Ok(result) => match result.status {
                UpdateStatus::Updated => Ok(()),
                UpdateStatus::NotUpdatedButExists => {
                    let record = result.found;

                    if record.operating_saga_id == None
                        && record.replacement_state
                            == RegionReplacementState::Running
                        && record.new_region_id == Some(new_region_id)
                        && record.delete_old_region_volume_id // <-------- here
                            == Some(old_region_volume_id) // <-------- and here
                    {
```

Can schema.rs be changed too? I don't think so, it seems like it has to map to the actual column name.

Reply:
Eh, it's fine. Just leave it. No need to go through this hassle right now. #5858 will likely clear up any miscommunication anyway.